Implement nxp USB device controllers (udc_mcux)

MarkWangChinese opened the pull request:
Create the mcux usb device controller driver (udc_mcux.c) based on NXP SDK USB Device controller drivers (usb_device_ehci.c, usb_device_ip3511.c etc).
Support EHCI and IP3511 controllers. Shell demo's transfer function is verified on RT1060, RT685 and LPC55S69.


Edit: addresses list found in #42066

tmon-nordic commented:
Same as with the host PR, the order of commits has to be bisectable, meaning first update hal and then use it, not the other way round.

MarkWangChinese commented:
> Same as with the host PR, the order of commits has to be bisectable, meaning first update hal and then use it, not the other way round.

Thanks @tmon-nordic , I have updated the commits order. Updating soc device tree firstly, enabling the udc/uhc driver secondly, then enabling the board level device tree and related clock. Please help to check whether it is OK.

dleach02 commented:
@jfischer-no, need your attention to this PR. 

@tmon-nordic, could you confirm the changes address your raised issue?

tmon-nordic commented:
> @tmon-nordic, could you confirm the changes address your raised issue?

Yes, the commit order seems fine now.

dleach02 commented:
@MarkWangChinese can you resolve the merge conflicts please.

dleach02 commented:
@jfischer-no can you provide review support?

MarkWangChinese commented:
> @MarkWangChinese can you resolve the merge conflicts please.

resolved.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,898 @@
+
+/* TODO: implement the cache maintenance
+ * solution1: Use the non-cached buf to do memcpy before/after giving buffer to usb controller.
+ * solution2: Use cache API to flush/invalid cache. but it needs the given buffer is
+ * cache line size aligned and the buffer range cover multiple of cache line size block.
+ * Need to change the usb stack to implement it, will try to implement it later.
```
Thanks Tomasz. It needs the usb device stack's support because the usb transfer buf is from usb stack or application, we can plan it for future.

dleach02 commented:
@hakehuang can you run HW test range USB tests for lpc55s69, rt1060, rt685. Report results to this PR so we can move on getting things merged.

hakehuang commented:
> @hakehuang can you run HW test range USB tests for lpc55s69, rt1060, rt685. Report results to this PR so we can move on getting things merged.

@MarkWangChinese  can you resolve the west.yaml conflicts, so that I can run test on your PR?

MarkWangChinese commented:
> > @hakehuang can you run HW test range USB tests for lpc55s69, rt1060, rt685. Report results to this PR so we can move on getting things merged.
> 
> @MarkWangChinese can you resolve the west.yaml conflicts, so that I can run test on your PR?

Thanks Hake, Resolved.

hakehuang commented:
@MarkWangChinese , build failure for mimxrt1060_evk

<html>
<body>
<!--StartFragment--><span><span class="ui-provider a b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak" dir="ltr">
CASE_NAME | ZEPHYR_VERSION | RESULT | PLATFORM | TEST_PATH
-- | -- | -- | -- | --
usb.desc_sections | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | tests/subsys/usb/desc_sections
usb.device | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | tests/subsys/usb/device
usb.bos | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | tests/subsys/usb/bos
sample.usb.mass_ram_none | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/mass
sample.usb.mass_ram_fat | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/mass
sample.usb.mass_sdhc_fatfs | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/mass
sample.usb.webusb | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/webusb
sample.usb.cdc-acm-composite | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/cdc_acm_composite
sample.usb.hid-mouse | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/hid-mouse
sample.usb.hid | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/hid
sample.usb.loopback | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/testusb
sample.usb.cdc-acm | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/cdc_acm
sample.usb.cdc-acm.comp | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/cdc_acm
sample.usb.audio.headset | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/audio/headset
sample.usb.audio.headphones_microphone | v3.6.0-3416-gbff06e7cc037 | Build failure | mimxrt1060_evk | samples/subsys/usb/audio/headphones_microphone

</span></span><!--EndFragment-->
</body>
</html>

MarkWangChinese commented:
> @MarkWangChinese , build failure for mimxrt1060_evk
> 
> error log
> 
> ```
> zephyr/drivers/usb/device/CMakeFiles/drivers__usb__device.dir/usb_dc_mcux.c.obj -c /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/drivers/usb/device/usb_dc_mcux.c
> In file included from /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/include/zephyr/toolchain.h:50,
>                  from /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/include/zephyr/sys/util.h:18,
>                  from /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/soc/nxp/imxrt/imxrt10xx/./soc.h:10,
>                  from /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/drivers/usb/device/usb_dc_mcux.c:8:
> /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/include/zephyr/toolchain/gcc.h:87:36: error: static assertion failed: "Only one USB device supported"
>    87 | #define BUILD_ASSERT(EXPR, MSG...) _Static_assert(EXPR, "" MSG)
>       |                                    ^~~~~~~~~~~~~~
> /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/modules/hal_nxp/usb/./usb_device_config.h:44:1: note: in expansion of macro 'BUILD_ASSERT'
>    44 | BUILD_ASSERT(NUM_INSTS <= 1, "Only one USB device supported");
>       | ^~~~~~~~~~~~
> /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/include/zephyr/toolchain/gcc.h:87:36: error: static assertion failed: "Only one USB device supported"
>    87 | #define BUILD_ASSERT(EXPR, MSG...) _Static_assert(EXPR, "" MSG)
>       |                                    ^~~~~~~~~~~~~~
> /home/jenkins/agent/workspace/bitbucket_build_platform/zephyr/drivers/usb/device/usb_dc_mcux.c:76:1: note: in expansion of macro 'BUILD_ASSERT'
>    76 | BUILD_ASSERT(NUM_INSTS <= 1, "Only one USB device supported");
>       | ^~~~~~~~~~~~
> [97/150] Building C object zephyr/drivers/timer/CMakeFiles/drivers__timer.dir/sys_clock_init.c.obj
> ```
> 
> CASE_NAME	ZEPHYR_VERSION	RESULT	PLATFORM	TEST_PATH
> usb.desc_sections	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	tests/subsys/usb/desc_sections
> usb.device	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	tests/subsys/usb/device
> usb.bos	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	tests/subsys/usb/bos
> sample.usb.mass_ram_none	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/mass
> sample.usb.mass_ram_fat	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/mass
> sample.usb.mass_sdhc_fatfs	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/mass
> sample.usb.webusb	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/webusb
> sample.usb.cdc-acm-composite	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/cdc_acm_composite
> sample.usb.hid-mouse	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/hid-mouse
> sample.usb.hid	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/hid
> sample.usb.loopback	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/testusb
> sample.usb.cdc-acm	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/cdc_acm
> sample.usb.cdc-acm.comp	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/cdc_acm
> sample.usb.audio.headset	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/audio/headset
> sample.usb.audio.headphones_microphone	v3.6.0-3416-gbff06e7cc037	Build failure	mimxrt1060_evk	samples/subsys/usb/audio/headphones_microphone

Thanks Hake, The old USB device stack only supports one controller, the new USB device stack supports multiple controllers. but they use the same DT. The two versions of USB device stack will coexist now. In order to fix the old stack's building error, I remove one controller node of mimxrt1060_evk's device tree.
In this PR, the main purpose is to support the new usb device stack, please help to pay attention to the samples that have `usbd_next_prj.conf` for new usb device stack.

hakehuang commented:
> Thanks Hake, The old USB device stack only supports one controller, the new USB device stack supports multiple controllers. but they use the same DT. The two versions of USB device stack will coexist now. In order to fix the old stack's building error, I remove one controller node of mimxrt1060_evk's device tree.
> In this PR, the main purpose is to support the new usb device stack, please help to pay attention to the samples that have `usbd_next_prj.conf` for new usb device stack.

in this case, can you enable the supported platform with the `usbd_next_prj.conf` samples application, now it is only tested on frdm_k64f.

MarkWangChinese commented:
> in this case, can you enable the supported platform with the `usbd_next_prj.conf` samples application, now it is only tested on frdm_k64f.

Hi Hake, The follow usb next device stack's samples' sample.yaml are updated, and I do basic test. You can test them on mimxrt685_evk/mimxrt685s/cm33, lpcxpresso55s69/lpc55s69/cpu0 and mimxrt1060_evk.

- samples\subsys\usb\cdc_acm\sample.usb_device_next.cdc-acm
- samples\subsys\usb\mass\sample.usb_device_next.mass_ram_none
- samples\subsys\usb\shell\sample.usbd.shell

And the follow test should be for all boards, don't need to change testcase.yaml.

- tests\subsys\usb\device_next\usb.device_next.build_all && usb.device_next

hakehuang commented:
> > in this case, can you enable the supported platform with the `usbd_next_prj.conf` samples application, now it is only tested on frdm_k64f.
> 
> Hi Hake, The follow usb next device stack's samples' sample.yaml are updated, and I do basic test. You can test them on mimxrt685_evk/mimxrt685s/cm33, lpcxpresso55s69/lpc55s69/cpu0 and mimxrt1060_evk.
> 
> * samples\subsys\usb\cdc_acm\sample.usb_device_next.cdc-acm
> * samples\subsys\usb\mass\sample.usb_device_next.mass_ram_none
> * samples\subsys\usb\shell\sample.usbd.shell
> 
> And the follow test should be for all boards, don't need to change testcase.yaml.
> 
> * tests\subsys\usb\device_next\usb.device_next.build_all && usb.device_next

test PASS for all supported platforms on usb_device mode

dleach02 commented:
@mmahadevan108 @jfischer-no @tmon-nordic 

Please provide a review so we can get this PR merged.

jfischer-no commented on drivers/usb/udc/Kconfig.mcux:
```diff
@@ -0,0 +1,38 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config USB_UDC_NXP_EHCI
+	bool "NXP MCUX USB EHCI Device controller driver"
+	default y
+	depends on DT_HAS_NXP_EHCI_ENABLED
+	select USB_UDC_NXP_MCUX
+	help
+	  NXP MCUX USB Device Controller Driver for EHCI.
+
+config USB_UDC_NXP_IP3511
+	bool "NXP MCUX USB IP3511 Device controller driver"
+	default y
+	depends on DT_HAS_NXP_LPCIP3511_ENABLED
+	select USB_UDC_NXP_MCUX
+	help
+	  NXP MCUX USB Device Controller Driver for KHCI.
```
I cannot find UDC driver implementations for controller EHCI and IP3511 in this PR. I think there is a misunderstanding. Is this a shim driver for a vendor USB middleware? Considering that Zephyr is there to fix all the issues with fancy vendor SDKs and provide a maintainable "infrastructure", this does not really look like a step forward. Can we please at least have a driver per controller type, that provides maintainable code for specific controller type and does not hide details in some other repository?

MarkWangChinese replied:
Hi @jfischer-no Yes, it is a shim driver that uses the NXP SDK USB controller drivers. The SDK USB controller drivers are very stable because they are tested with many NXP SDK USB stack examples that are released many times in NXP SDK, so I prefer to use them. I have did some dependency simplification of NXP SDK USB controller drivers (https://github.com/zephyrproject-rtos/hal_nxp/pull/353) to use it more easily in Zephyr.
Because the SDK USB controller drivers provide the same interface, so I only create one udc_mcux.c for EHCI and IP3511. I can create two (udc_mcux_ehci,c and udc_mcux_ip3511.c) for them, but they will be almost same. I am OK if two is more  consistent with Zephyr style. Please let me know your thought.

MarkWangChinese commented on dts/bindings/usb/nxp,usbphy.yaml:
```diff
@@ -0,0 +1,30 @@
+# Copyright  2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: NXP USB High Speed Phy
```
NXP MCU only supports one type of USB PHY that needs software driver. It is used in RTxxxx, RTxxx, MCX and Kinetis platforms. I will add more descriptions.

MarkWangChinese replied:
Updated

dleach02 commented on dts/bindings/usb/nxp,usbphy.yaml:
```diff
@@ -0,0 +1,30 @@
+include: base.yaml
+
+properties:
+  d-cal:
+    type: int
+    required: true
+    description: |
+      It is board level's value that is used to trim the nominal 17.78mA
+      current source for the High Speed TX drivers on USB_DP and USB_DM.
+
+  txcal45dp:
+    type: int
+    required: true
+    description: |
+      It is board level's value that is used to trim the nominal 17.78mA
+      current source for the High Speed TX drivers on USB_DP and USB_DM.
+
+  txcal45dm:
+    type: int
+    required: true
+    description: |
+      It is board level's value that is used to trim the nominal 17.78mA
+      current source for the High Speed TX drivers on USB_DP and USB_DM.
```
Good catch.

@MarkWangChinese what @jfischer-no is pointing out is that there is a style guide for property names. Zephyr documentation is not really clear about this but we follow the https://docs.kernel.org/devicetree/bindings/dts-coding-style.html where node names use:

- Lowercase characters: [a-z]
- Digits: [0-9]
- Dash: -

The dash is used to separate out the terms. Please make the appropriate changes so we can get through this PR.

MarkWangChinese replied:
Thanks Johann and David. I have updated it.

mmahadevan108 commented:
@MarkWangChinese, please update the `west.yml`

MarkWangChinese commented:
> @MarkWangChinese, please update the `west.yml`

Thanks Mahesh, Updated.

dleach02 commented:
@jfischer-no can you revisit the PR

dleach02 commented:
Did a rebase to address west.yml

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+#endif
+
+#ifdef CONFIG_USB_UDC_NXP_EHCI
+#if defined(USBHS_STACK_BASE_ADDRS)
+	uint32_t usb_base_addrs[] = USBHS_STACK_BASE_ADDRS;
+#else
+	uint32_t usb_base_addrs[] = USBHS_BASE_ADDRS;
+#endif
+#endif
+#ifdef CONFIG_USB_UDC_NXP_IP3511
+#ifdef USB_BASE_ADDRS
+	uint32_t ip3511_fs_base[] = USB_BASE_ADDRS;
+#endif
+#ifdef USBHSD_BASE_ADDRS
+	uint32_t ip3511_hs_base[] = USBHSD_BASE_ADDRS;
+#endif
```
Use DT_INST_REG_ADDR to obtain base.

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	}
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+	if (config->need_to_init_phy) {
```
```c

	if (config->need_to_init_phy) {
		usb_phy_config_struct_t phy_config;

```

MarkWangChinese replied:
Updated

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+		phy_config.D_CAL = config->D_CAL;
+		phy_config.TXCAL45DP = config->TXCAL45DP;
+		phy_config.TXCAL45DM = config->TXCAL45DM;
+		USB_EhciPhyInit(priv->controller_id, 0u, &phy_config);
```
Why is `usb_phy_config_struct_t phy_config;` not a part of `udc_mcux_config`?

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+
+	data->caps.rwup = true;
+	data->caps.mps0 = USB_MCUX_MPS0;
+	priv->dev = dev;
```
No high speed capable controllers?

MarkWangChinese replied:
Added.

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+#define UDC_MCUX_PHY_CONFIGURE(n)							\
+	.need_to_init_phy = DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle) ? 1u : 0u,	\
+	.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-d-cal, 0),		\
+	.TXCAL45DP = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dp-ohms, 0),	\
+	.TXCAL45DM = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dm-ohms, 0),
```
This can be improved, I will comment on it once you have addressed other comments.

jfischer-no commented on samples/subsys/usb/cdc_acm/sample.yaml:
```diff
@@ -18,6 +18,9 @@ tests:
       - nrf52840dk/nrf52840
       - frdm_k64f
       - 96b_carbon/stm32f401xe
+      - mimxrt685_evk/mimxrt685s/cm33
+      - lpcxpresso55s69/lpc55s69/cpu0
+      - mimxrt1060_evk
```
Please only one entry per controller type. I guess mimxrt1060_evk uses the same controller type as mimxrt685_evk/mimxrt685s/cm33?

MarkWangChinese replied:
mimxrt1060_evk uses the ehci and the phy ip; lpc55s69 and mimxrt685s uses the same IP3511 and the phy ip. I will remove one later.

jfischer-no commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+#if defined(CONFIG_NOCACHE_MEMORY)
+			udc_mcux_nocache_free(mcux_msg->buffer);
+#endif
+			err = udc_submit_ep_event(dev, buf, 0);
```
Perhaps you need to check ZLP flag(`udc_ep_buf_has_zlp(buf)`) and add ZLP transaction somewhere here.

MarkWangChinese committed:
dts: arm: nxp: enable udc DT on lpc55s69, rt1060 and rt685

define usbphy in DT and controller DT node ref to usbphy node.
define the usbphy yaml and update ehci and ip3511 yaml for usbphy.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
drivers: udc: implement udc_mcux_ehci and udc_mcux_ip3511

udc_mcux_ehci is based on the MCUX USB controller driver
(usb_device_ehci.c); udc_mcux_ip3511 is based on the
MCUX USB controller driver (usb_device_lpcip3511.c);
add related Kconfig and CMake; include the usb_phy.h path in
modules/hal_nxp/usb/CMakeLists.txt because udc_mcux.c use it;
add related macros to usb_device_config.h;
update CMakeLists for udc_mcux_ehci and udc_mcux_ip3511.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
boards: nxp: enable mcux udc on lpc55s69, rt1060 and rt685

set DT node as Okay in board device tree;
add board level's d-cal, txcal45dp and txcal45dm to usbphy node;
enable usb clock; set USB_STACK_USE_DEDICATED_RAM for lpc55s69 and rt685;
load usb.ld for lpc55s69 and rt685.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
samples: usb: enable next usb device stack samples on NXP platforms

enable next usb device stack samples (cdc_acm, mass and shell)
on rt685 and mimxrt1060_evk

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented on drivers/usb/udc/CMakeLists.txt:
```diff
@@ -12,3 +12,4 @@ zephyr_library_sources_ifdef(CONFIG_UDC_KINETIS udc_kinetis.c)
 zephyr_library_sources_ifdef(CONFIG_UDC_SKELETON udc_skeleton.c)
 zephyr_library_sources_ifdef(CONFIG_UDC_VIRTUAL udc_virtual.c)
 zephyr_library_sources_ifdef(CONFIG_UDC_STM32 udc_stm32.c)
+zephyr_library_sources_ifdef(CONFIG_USB_UDC_NXP_MCUX udc_mcux.c)
```
Update the commit messages, please help to check.
Yes, EHCI is Enhanced Host Controller Interface. NXP extend it with usb device function too. So, the controller supports both usb host and usb device function.

MarkWangChinese commented on drivers/usb/udc/Kconfig.mcux:
```diff
@@ -0,0 +1,38 @@
+	help
+	  NXP MCUX USB Device Controller Driver for EHCI.
+
+config USB_UDC_NXP_IP3511
```
Hi Johann, I want to keep both `USB_UDC_NXP_EHCI` and `USB_UDC_NXP_IP3511` because it is used in other places. For example: the soc.c use it to initialize the usb clock as follow:
```
#if DT_NODE_HAS_STATUS(DT_NODELABEL(usb1), okay) && \
	(CONFIG_USB_DC_NXP_EHCI || CONFIG_USB_UDC_NXP_EHCI)
	CLOCK_EnableUsbhs0PhyPllClock(kCLOCK_Usb480M,
		DT_PROP_BY_PHANDLE(DT_NODELABEL(usb1), clocks, clock_frequency));
	CLOCK_EnableUsbhs0Clock(kCLOCK_Usb480M,
		DT_PROP_BY_PHANDLE(DT_NODELABEL(usb1), clocks, clock_frequency));
#if DT_NODE_HAS_STATUS(DT_NODELABEL(usb1), okay) && CONFIG_USB_DC_NXP_EHCI
	USB_EhciPhyInit(kUSB_ControllerEhci0, CPU_XTAL_CLK_HZ, &usbPhyConfig);
#endif
#endif
```
I have split the udc_mcux.c as udc_mcux_ehci.c and udc_mcux_ip3511.c in order to keep the  both `USB_UDC_NXP_EHCI` and `USB_UDC_NXP_IP3511` . Please share your comments that I should use one udc_mcux.c or use two (udc_mcux_ehci.c and udc_mcux_ip3511.c).

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+#include "usb.h"
+#include "usb_device_config.h"
+#include "usb_device_mcux_drv_port.h"
+#ifdef CONFIG_USB_UDC_NXP_PHY
```
NXP has multiple platforms like: RTxxx, RTxxxx, LPC, RW and Kinetis. RTxxxx uses the EHCI and the PHY ip; RTxxx and some LPC SoC uses IP3511 and the phy ip; Some LPC only use the IP3511 ip not the PHY ip; RW uses the EHCI IP not the PHY ip. So, the codes need to use the `CONFIG_USB_UDC_NXP_PHY` to determine whether to include the mcux phy driver header file.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+#define USBFSOTG_OUT_TOKEN	0x01
+
+#define USBFSOTG_PERID		0x04
+#define USBFSOTG_REV		0x33
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,
+	USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl
+};
+#endif
```
Have splite as two as above comments, so there is no this issue.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	usb_status_t status = kStatus_USB_Success;
+	uint8_t *data;
+	uint32_t len;
+	int current_sr;
```
Updated

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+		irq_unlock(current_sr);
+
+		if (USB_EP_DIR_IS_OUT(cfg->addr)) {
+			len = net_buf_tailroom(buf); /* MIN(net_buf_tailroom(buf), cfg->mps); */
```
Deleted.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	if (feed_buf) {
+		int ret = usb_mcux_ep_feed(dev, cfg, feed_buf);
+
+		return (((ret == -EBUSY) || (ret == -EACCES) || (ret == 0)) ? 0 : -EIO);
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+ * @param message   The device callback message handle.
+ *
+ * @return A USB error code or kStatus_USB_Success.
+ */
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+ */
+usb_status_t USB_DeviceNotificationTrigger(void *handle, void *msg)
+{
+	usb_device_callback_message_struct_t *mcux_msg =
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	case kUSB_DeviceNotifyAttach:
+		udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);
+		break;
+#endif
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+			usb_mcux_ep_try_feed(dev, udc_get_ep_cfg(dev, ep));
+		}
+		break;
+	}
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+
+	USB_DeviceLpcIp3511IsrFunction((void *)(&priv->device_instance));
+}
+#endif
```
Have splite as two as above comments, so it is similar with udc_kinetis.c now.

MarkWangChinese commented on drivers/usb/udc/udc_mcux.c:
```diff
@@ -0,0 +1,903 @@
+	}
+
+	current_sr = irq_lock();
+	if (!udc_ep_is_busy(dev, cfg->addr)) {
```
I am using the `busy `state to determine whether to prime next transfer to controller. I think there are two contexts that will prime transfers to controller. (1) `USB_DeviceNotificationTrigger` is called in isr when one transfer is done, then it will check queued transfer and prime it to controller; (2) application or other threads can queue transfer and prime it to controller too. So I need the `irq_lock` because I don't use the `work` as udc_kinetis.c did.

dleach02 commented:
@MarkWangChinese please address the merge conflicts
@jfischer-no please help review the update so we can guide this PR to merge

MarkWangChinese commented:
> @MarkWangChinese please address the merge conflicts @jfischer-no please help review the update so we can guide this PR to merge

conflict resolved.

jfischer-no commented:
@MarkWangChinese I tried it on IMXRT1050 board. MCUX shimd driver misses udc_device_speed() implementation and stack selects wrong configuration.
```
[2256087.269503] usb 3-2: config 1 interface 1 altsetting 0 bulk endpoint 0x82 has invalid maxpacket 64
[2256087.269510] usb 3-2: config 1 interface 1 altsetting 0 bulk endpoint 0x1 has invalid maxpacket 64
[2256087.269960] usb 3-2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.06
[2256087.269965] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[2256087.269969] usb 3-2: Product: USBD CDC ACM sample
[2256087.269972] usb 3-2: Manufacturer: Zephyr Project
[2256087.269974] usb 3-2: SerialNumber: 500100B42B3529D2
```

Please implement `.device_speed`, check udc_dwc2.c for reference. After the reset event, the stack checks the actual speed of the device to select the correct speed configuration. Should also be fixed for IP3511 if it supports high speed controllers. It would be nice to have another high speed controller supported.

(By the way, some bad actors changed the default runner to a "LinkServer" after many years of having "vendor neutral" runners for NXP boards. This is very annoying and should not be allowed in a project like Zephyr. I really don't want another redundant software/dependency on my workstation!)

dleach02 commented:
> @MarkWangChinese I tried it on IMXRT1050 board. MCUX shimd driver misses udc_device_speed() implementation and stack selects wrong configuration.
> 

@jfischer-no is there a USB test case in the tree that would have exercised this?

jfischer-no commented:
> > @MarkWangChinese I tried it on IMXRT1050 board. MCUX shimd driver misses udc_device_speed() implementation and stack selects wrong configuration.
> 
> @jfischer-no is there a USB test case in the tree that would have exercised this?

We have nothing to test exactly that now, but in the future it would be possible, e.g. for the MCUX shim driver, IMXRT1050 has two high speed phys that could be used to create a test loop.

But this one issue is pretty obvious.
Also, you can run USB2CV on all controller variants(recommended, it may help to find other problems in the driver), a high speed controller using MCUX shim driver is not compliant and would fail in USB2CV (Endpoint Descriptor Test, Other Speed Endpoint Descriptor Test).

MarkWangChinese commented:
> @MarkWangChinese I tried it on IMXRT1050 board. MCUX shimd driver misses udc_device_speed() implementation and stack selects wrong configuration.
> 
> ```
> [2256087.269503] usb 3-2: config 1 interface 1 altsetting 0 bulk endpoint 0x82 has invalid maxpacket 64
> [2256087.269510] usb 3-2: config 1 interface 1 altsetting 0 bulk endpoint 0x1 has invalid maxpacket 64
> [2256087.269960] usb 3-2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.06
> [2256087.269965] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [2256087.269969] usb 3-2: Product: USBD CDC ACM sample
> [2256087.269972] usb 3-2: Manufacturer: Zephyr Project
> [2256087.269974] usb 3-2: SerialNumber: 500100B42B3529D2
> ```
> 
> Please implement `.device_speed`, check udc_dwc2.c for reference. After the reset event, the stack checks the actual speed of the device to select the correct speed configuration. Should also be fixed for IP3511 if it supports high speed controllers. It would be nice to have another high speed controller supported.
> 
> (By the way, some bad actors changed the default runner to a "LinkServer" after many years of having "vendor neutral" runners for NXP boards. This is very annoying and should not be allowed in a project like Zephyr. I really don't want another redundant software/dependency on my workstation!)

Implemented the `.device_speed`

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+											\
+	static struct udc_mcux_config priv_config_##n = {				\
+		UDC_MCUX_PHY_CONFIGURE(n)						\
+		.base = (uint32_t)DT_INST_REG_ADDR(n),					\
```
`		.base = (uintptr_t)DT_INST_REG_ADDR(n),					\` ?

MarkWangChinese replied:
Change `base` type as uintptr_t

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+}
+
+static const struct udc_api usb_mcux_api = {
+	.device_speed = udc_mcux_device_speed,
+	.ep_enqueue = usb_mcux_ep_enqueue,
+	.ep_dequeue = usb_mcux_ep_dequeue,
+	.ep_set_halt = usb_mcux_ep_set_halt,
+	.ep_clear_halt = usb_mcux_ep_clear_halt,
+	.ep_try_config = NULL,
+	.ep_enable = usb_mcux_ep_enable,
+	.ep_disable = usb_mcux_ep_disable,
+	.host_wakeup = usb_mcux_host_wakeup,
+	.set_address = usb_mcux_set_address,
+	.enable = usb_mcux_enable,
+	.disable = usb_mcux_disable,
+	.init = usb_mcux_init,
+	.shutdown = usb_mcux_shutdown,
+	.lock = usb_mcux_lock,
+	.unlock = usb_mcux_unlock,
```
Should the prefix be usb_mcux? I suggest to prefix it consistently with udc_mcux.

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/udc/Kconfig.mcux:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config USB_UDC_NXP_EHCI
```
Prefix must be UDC_, not USB_UDC. Please fix it for all new Kconfig options.
config UDC_NXP_EHCI

MarkWangChinese replied:
Updated. Hi @dleach02 Because this change, I create one hal_nxp pr https://github.com/zephyrproject-rtos/hal_nxp/pull/400

jfischer-no replied:
Can you do
```
  include_ifdef(CONFIG_DT_HAS_NXP_USBPHY_ENABLED  middleware_usb_phy)
  include_ifdef(CONFIG_UDC_NXP_EHCI               middleware_usb_device_ehci)
  include_ifdef(CONFIG_UDC_NXP_IP3511             middleware_usb_device_ip3511fs)
```
in `modules/hal_nxp/CMakeLists.txt`?

dleach02 replied:
> Updated. Hi @dleach02 Because this change, I create one hal_nxp pr [zephyrproject-rtos/hal_nxp#400](https://github.com/zephyrproject-rtos/hal_nxp/pull/400)

Thanks

jfischer-no commented on drivers/usb/udc/Kconfig.mcux:
```diff
@@ -0,0 +1,24 @@
+	help
+	  NXP MCUX USB Device Controller Driver for KHCI.
+
+config USB_UDC_NXP_PHY
+	bool "NXP MCUX USB Phy"
+	default y
+	depends on DT_HAS_NXP_USBPHY_ENABLED
+	help
+	  NXP MCUX USB Phy.
```
I see the only justification for this is use in `modules/hal_nxp/usb/CMakeLists.txt`. Can we use DT_HAS_NXP_USBPHY_ENABLED in `modules/hal_nxp/usb/CMakeLists.txt` there directly?

MarkWangChinese replied:
Remove `COFNIG_USB_UDC_NXP_PHY` and use `CONFIG_DT_HAS_NXP_USBPHY_ENABLED `

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+	memcpy(buf->data, setup, 8);
+	net_buf_add(buf, 8);
+
+	if ((setup->RequestType.type == USB_REQTYPE_TYPE_STANDARD) &&
+		(!setup->wLength) &&
+		(!(((setup->bmRequestType) >> 7) & 0x01U)) &&
```
Please write it consistently, if you start with `setup->RequestType.type == USB_REQTYPE_TYPE_STANDARD`, then 

```c
	if (setup->RequestType.type == USB_REQTYPE_TYPE_STANDARD &&
	    setup->RequestType.direction == USB_REQTYPE_DIR_TO_DEVICE &&
	    setup->bRequest == USB_SREQ_SET_ADDRESS &&
	    setup->wLength == 0) {
```

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+		}
+		udc_submit_event(dev, UDC_EVT_RESET, 0);
+		break;
+#if (defined(USB_DEVICE_CONFIG_ERROR_HANDLING) && (USB_DEVICE_CONFIG_ERROR_HANDLING > 0U))
+	case kUSB_DeviceNotifyError:
+		udc_submit_event(dev, UDC_EVT_ERROR, -EIO);
+		break;
+#endif
+#if (defined(USB_DEVICE_CONFIG_LOW_POWER_MODE) && (USB_DEVICE_CONFIG_LOW_POWER_MODE > 0U))
+	case kUSB_DeviceNotifySuspend:
+		udc_set_suspended(dev, true);
+		udc_submit_event(dev, UDC_EVT_SUSPEND, 0);
+		break;
+	case kUSB_DeviceNotifyResume:
+		udc_set_suspended(dev, false);
+		udc_submit_event(dev, UDC_EVT_RESUME, 0);
+		break;
+#if (defined(USB_DEVICE_CONFIG_LPM_L1) && (USB_DEVICE_CONFIG_LPM_L1 > 0U))
+	case kUSB_DeviceNotifyLPMSleep:
+		break;
+#endif
+#endif
+#if (defined(USB_DEVICE_CONFIG_DETACH_ENABLE) && (USB_DEVICE_CONFIG_DETACH_ENABLE > 0U))
+	case kUSB_DeviceNotifyDetach:
+		udc_submit_event(dev, UDC_EVT_VBUS_REMOVED, 0);
+		break;
+	case kUSB_DeviceNotifyAttach:
+		udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);
+		break;
+#endif
```
I looked at the definition of usb_device_notification_t and there is no reason for all the `#if-#endif`s. Please clean it up.

jfischer-no reviewed (commented):
I looked at udc_mcux_ehci.c, but most of the comments apply to udc_mcux_ip3511.c as well.

MarkWangChinese commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+static const usb_device_controller_interface_struct_t udc_mcux_if = {
+	USB_DeviceEhciInit, USB_DeviceEhciDeinit, USB_DeviceEhciSend,
+	USB_DeviceEhciRecv, USB_DeviceEhciCancel, USB_DeviceEhciControl
+};
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+
+	/* get the right controller id */
+	priv->controller_id = 0xFFu; /* invalid value */
+	for (index = 0; index < ARRAY_SIZE(usb_base_addrs); index++) {
```
Updated.

MarkWangChinese commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,920 @@
+	/* MCUX USB controller drivers use one ID to tell controller
+	 * drivers which controller is used. This part of codes are
+	 * convert the IP base address to the ID value.
+	 */
```
Updated.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,918 @@
+#include "usb_device_config.h"
+#include "usb_device_mcux_drv_port.h"
+#include "usb_device_ehci.h"
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
```
Please remove `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED`, it should not be necessary.

MarkWangChinese replied:
As I explained, NXP has may platforms. Even the same USB controller IP is used, the PHY is used or not in different platforms. For example: (1) LPC54608 uses the IP3511 driver, the phy is not used; LPC55S69 uses the IP3511 driver, the phy is used. (2) RW610 uses the EHCI driver, the phy is not used; RTxxxx series uses the EHCI driver, the phy is used.
So, the MACRO is needed here.

jfischer-no replied:
I do not care how many platforms NXP has. There is no need to include this header conditionally.

decsny replied:
why does it matter?

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,918 @@
+	struct udc_ep_config *ep_cfg_out;
+	uintptr_t base;
+	const struct pinctrl_dev_config *pincfg;
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+	usb_phy_config_struct_t phy_config;
+	uint8_t need_to_init_phy;
+#endif
```
You can simplify it to just
```c
	usb_phy_config_struct_t *phy_config;
```
Code that always builds is preferable, you also do not break multiple instance support with/without phy (no idea if it even exists.).

MarkWangChinese replied:
Change to as only `usb_phy_config_struct_t *phy_config`

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,918 @@
+		return -ENOMEM;
+	}
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+	if (config->need_to_init_phy) {
+		USB_EhciPhyInit(priv->controller_id, 0u,
+			(usb_phy_config_struct_t *)&config->phy_config);
+	}
+#endif
```
Can be
```c
	if (config->phy_config != NULL) {
		USB_EhciPhyInit(priv->controller_id, 0u, config->phy_config);
	}
```

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,918 @@
+	USB_DeviceEhciRecv, USB_DeviceEhciCancel, USB_DeviceEhciControl
+};
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+#define UDC_MCUX_PHY_CONFIGURE(n)							\
+	.need_to_init_phy = DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle) ? 1u : 0u,	\
+	.phy_config.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-d-cal, 0),	\
+	.phy_config.TXCAL45DP =								\
+		DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dp-ohms, 0),	\
+	.phy_config.TXCAL45DM =								\
+		DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dm-ohms, 0),
+#else
+#define UDC_MCUX_PHY_CONFIGURE(n)
+#endif
+
+#define USB_MCUX_EHCI_DEVICE_DEFINE(n)							\
```
Now you can use `COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle), ...) ` to define phy_config_##n and assign it to the pointer phy_config. I pushed all the changes to my branch, you can pull and squash it, and maybe adapt ip3511 driver accordingly. See https://github.com/jfischer-no/zephyr/commit/7fe0c43e6c1bce73bdded3ef073a4b7f026ad72b (https://github.com/jfischer-no/zephyr/commit/7fe0c43e6c1bce73bdded3ef073a4b7f026ad72b.patch)

MarkWangChinese replied:
I have combine your changes with my side's codes because `phy_config` is not needed in the platforms that doesn't have the PHY. Please help to review.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,927 @@
+	USB_DeviceEhciRecv, USB_DeviceEhciCancel, USB_DeviceEhciControl
+};
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+#define UDC_MCUX_PHY_DEFINE(n)								\
+static usb_phy_config_struct_t phy_config_##n = {					\
+	.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_d_cal, 0),		\
+	.TXCAL45DP = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_cal_45_dp_ohms, 0),	\
+	.TXCAL45DM = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_cal_45_dm_ohms, 0),	\
+}
+
+#define UDC_MCUX_PHY_DEFINE_OR(n)							\
+	COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle),			\
+		    (UDC_MCUX_PHY_DEFINE(n)), ())
+
+#define UDC_MCUX_PHY_CFG_PTR_OR_NULL(n)							\
+	.phy_config = COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle),		\
+		    (&phy_config_##n), (NULL))
+#else
+#define UDC_MCUX_PHY_DEFINE_OR(n)
+#define UDC_MCUX_PHY_CFG_PTR_OR_NULL(n)
+#endif
```
Not addressed, `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED` is not required.

MarkWangChinese replied:
I don't understand. In one platform that doesn't have the PHY driver, so the usb_phy.h is not present/valid in this platform. Why should I include this file?
I use `CONFIG_DT_HAS_NXP_USBPHY_ENABLED` to reduce the `phy_config` here because it is not needed if the platform doesn't have this phy.

MarkWangChinese replied:
I can remove all the phy related codes from the driver and put it in the soc.c of different platforms. The previous version NXP usb device controller driver (usb_dc_mcux.c) works this way. 
Please let me know your thought.

jfischer-no replied:
The phy initialization code is fine and belongs to the shim driver. There is nothing in [phy/usb_phy.h](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/middleware/mcux-sdk-middleware-usb/phy/usb_phy.h) that requires you to include it conditionally, it can always build. Therefore, all your `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED` can be removed, resulting in much cleaner code without unnecessary #ifdef. Here it is doubly unnecessary, whatever phy is present and needs to be initialized is covered by the macros above.

MarkWangChinese replied:
Hi Johann, In SDK USB stack, I cannot guarantee the usb_phy.h will not contain the soc/platform/IP level codes in future even it only contains the common C codes now. The reason is: NXP releases SDK package by platform, one platform has one package. usb_phy.h is only present in the package of platforms that have this PHY, and it is only valid/used in the platform, it is never built with the platforms that doesn't have the PHY. Without NXP SDK release's pre-building, I cannot guarantee usb_phy.h will build successfully with the platforms that doesn't have this phy forever.

jfischer-no replied:
@MarkWangChinese Got it, can you please add a comment before `#include "usb_phy.h`" that the header is not available for every package, thanks.

But, `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED ... #else` is not necessary here, see https://github.com/jfischer-no/zephyr/commit/7652d1647de7dbf1cdce4b6f706605f6ea8c9e48 (https://github.com/jfischer-no/zephyr/commit/7652d1647de7dbf1cdce4b6f706605f6ea8c9e48.patch)

dleach02 replied:
Dev Review discussion:
@MarkWangChinese we discussed this PR with @jfischer-no in today's dev review meeting. He is would like to see the #ifdef removed and will approve the PR for merging. If we find later on subsequent platforms this is a problem we can address then.

Johann also re-kicked the CI so that is clean. So we will be able to merge this PR on Friday 6/14.

jfischer-no replied:
@dleach02 @MarkWangChinese Actually let get it in, we have CI green and do not know what happens tomorrow :smiley:, https://github.com/zephyrproject-rtos/zephyr/pull/70401#discussion_r1638374321 can be follow up.

MarkWangChinese replied:
I have created one PR https://github.com/zephyrproject-rtos/zephyr/pull/74382, @jfischer-no please help to review.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,927 @@
+	err = udc_mcux_control(dev, kUSB_DeviceControlGetSpeed, &mcux_speed);
+	if (err) {
+		/* In current all NXP USB device drivers,
+		 * it will not reutrn err if parameter is right
```
return?

Should the comment be:
```
/*
 * In the current version of all NXP USB device drivers,
 * no error is returned if the parameter is correct.
 */
```
?

Similar in (udc_mcux_ip3511.c)

MarkWangChinese replied:
Update the comments.

jfischer-no commented on drivers/usb/udc/udc_mcux_ip3511.c:
```diff
@@ -0,0 +1,888 @@
+		}
+	}
+
+	data->caps.rwup = true;
```
```c
/* Requires udc_mcux_host_wakeup() implementation */
data->caps.rwup = false;
```

MarkWangChinese replied:
Updated. Will check suspend feature later with new PR.

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -0,0 +1,927 @@
+		}
+	}
+
+	data->caps.rwup = true;
```
```c
/* Requires udc_mcux_host_wakeup() implementation */
data->caps.rwup = false;
```

MarkWangChinese replied:
Updated. Will check suspend feature later with new PR.

dleach02 commented:
The CI issues are unrelated. A fix has been posted by the TF-M folks [#74052]
